### ISSUE TITLE
Android AppBundle issue with 1.10.7 .. 1.10.14

### ISSUE THREAD

**Issue by PieterAelse** [flutter/flutter#43347]
## Steps to Reproduce
1. Update Flutter from 1.9.1... to 1.10.7 or higher(*)
2. Run flutter build appbundle
3. Install appbundle on any Android phone
4. App crashes at startup

* I didn't test with 1.10.x lower than 1.10.7.

## Logs
Android LogCat doesn't show any useful stacktrace, but this is what is shown after clicking and the crash:

> 2019-10-23 16:29:53.051 1301-7111/? I/ActivityTaskManager: START u0 {act=android.intent.action.MAIN cat=[android.intent.category.LAUNCHER] flg=0x10200000 cmp=com.signify.hue.blue/com.signify.hue.hueblueapp.MainActivity bnds=[641,1647][843,1917]} from uid 10096
> 2019-10-23 16:29:53.083 1301-1345/? I/ActivityManager: Start proc 29090:com.signify.hue.blue/u0a290 for activity {com.signify.hue.blue/com.signify.hue.hueblueapp.MainActivity}
> 2019-10-23 16:29:53.115 29090-29090/? I/chatty: uid=10290(com.signify.hue.blue) identical 2 lines
> 2019-10-23 16:29:53.218 29090-29115/? I/FA: To enable faster debug mode event logging run:
>       adb shell setprop debug.firebase.analytics.app com.signify.hue.blue
> 2019-10-23 16:29:53.342 29090-29090/? D/FirebasePerformance: onResume(): com.signify.hue.hueblueapp.MainActivity: 219126 microseconds
> 2019-10-23 16:29:53.498 1301-1343/? I/ActivityTaskManager: Displayed com.signify.hue.blue/com.signify.hue.hueblueapp.MainActivity: +444ms
> 2019-10-23 16:29:53.562 29210-29210/? A/DEBUG: pid: 29090, tid: 29090, name: ignify.hue.blue  >>> com.signify.hue.blue <<<
> 2019-10-23 16:29:53.584 630-630/? E/Layer: [Surface(name=AppWindowToken{dbb147d token=Token{76a7ad4 ActivityRecord{36eb827 u0 com.signify.hue.blue/com.signify.hue.hueblueapp.MainActivity t828}}})/@0xf0e6970 - animation-leash#0] No local sync point found
> 2019-10-23 16:29:53.632 29210-29210/? A/DEBUG:       #01 pc 000000000001cfa0  /data/app/com.signify.hue.blue-jPD_qly-cXX1NtbVUo5z6A==/split_config.arm64_v8a.apk
> 2019-10-23 16:29:53.989 1301-29218/? W/ActivityTaskManager:   Force finishing activity com.signify.hue.blue/com.signify.hue.hueblueapp.MainActivity
> 2019-10-23 16:29:54.044 1301-1741/? W/InputDispatcher: channel '1d1cb91 com.signify.hue.blue/com.signify.hue.hueblueapp.MainActivity (server)' ~ Consumer closed input channel or an error occurred.  events=0x9
> 2019-10-23 16:29:54.044 1301-1741/? E/InputDispatcher: channel '1d1cb91 com.signify.hue.blue/com.signify.hue.hueblueapp.MainActivity (server)' ~ Channel is unrecoverably broken and will be disposed!
> 2019-10-23 16:29:54.046 1301-7048/? I/ActivityManager: Process com.signify.hue.blue (pid 29090) has died: vis+99 TOP 
> 2019-10-23 16:29:54.046 1301-5810/? I/WindowManager: WIN DEATH: Window{1d1cb91 u0 com.signify.hue.blue/com.signify.hue.hueblueapp.MainActivity}
> 2019-10-23 16:29:54.046 1301-5810/? W/InputDispatcher: Attempted to unregister already unregistered input channel '1d1cb91 com.signify.hue.blue/com.signify.hue.hueblueapp.MainActivity (server)'
> 2019-10-23 16:29:54.493 1301-1338/? W/ActivityTaskManager: Activity top resumed state loss timeout for ActivityRecord{36eb827 u0 com.signify.hue.blue/com.signify.hue.hueblueapp.MainActivity t-1 f}
> 
> 

## Flutter analyze

> [✓] Flutter (Channel unknown, v1.10.7, on Mac OS X 10.15 19A602, locale en-GB)
>     • Flutter version 1.10.7 at /[path]/flutter
>     • Framework revision e70236e36c (3 weeks ago), 2019-10-02 09:32:30 -0700
>     • Engine revision 9e6314d348
>     • Dart version 2.6.0 (build 2.6.0-dev.0.0 1103600280)
> 
> [✓] Android toolchain - develop for Android devices (Android SDK version 29.0.1)
>     • Android SDK at /Users/pieter/Library/Android/sdk
>     • Android NDK location not configured (optional; useful for native profiling support)
>     • Platform android-29, build-tools 29.0.1
>     • ANDROID_HOME = /Users/pieter/Library/Android/sdk
>     • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
>     • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)
>     • All Android licenses accepted.
> 
> [✓] Xcode - develop for iOS and macOS (Xcode 11.1)
>     • Xcode at /Applications/Xcode.app/Contents/Developer
>     • Xcode 11.1, Build version 11A1027
>     • CocoaPods version 1.8.4
> 
> [✓] Android Studio (version 3.5)
>     • Android Studio at /Applications/Android Studio.app/Contents
>     • Flutter plugin version 40.0.2
>     • Dart plugin version 191.8423
>     • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)
> 
> [✓] IntelliJ IDEA Community Edition (version 2019.2.3)
>     • IntelliJ at /Applications/IntelliJ IDEA CE.app
>     • Flutter plugin version 40.2.4
>     • Dart plugin version 192.7402
> 
> [✓] Connected device (1 available)
>     • Pixel 2 • FA7BF1A00613 • android-arm64 • Android 10 (API 29)
> 
> • No issues found!

**Comment by xster** [comment]
Can we get a bit more details with the repro of this issue?

Is this a standard `flutter create`ed app or add-to-app? Which plugins were involved?

**Comment by blasten** [comment]
As @xster said, would you mind listing the plugins and their versions?

Also, how did you install the app bundle? What it from the Play Store or did you use the extraction tool and `adb install`?

**Comment by PieterAelse** [comment]
I've looked into this issue the last days and in the end I was able to fix it by adding Proguard rules to a package/plugin we created ourselves. So the stacktrace @remonh87 mentioned was because of some class(es) getting stripped out with Proguard.

It turns out that starting Flutter 1.10.x when doing `flutter build appbundle` the `gradle.properties` file in the Android project gets edited, the line `android.enableR8=true` is added. This enables Proguard with some optimizations/minification/shrinking. 
In order to fix the stacktrace above I had to add some specific `-keep` rules to `proguard-rules.pro` so make sure things in our plugin stay intact.

Should it be mentioned somewhere that R8 gets enabled by Flutter when building a bundle? This could be confusing.

**Comment by PieterAelse** [comment]
To elaborate a bit more:

At first I thought the problem was with some Firebase plugins. So I've created a new project and added the following plugins:
```
firebase_core: "^0.4.0+9"
firebase_crashlytics: "^0.1.0+3"
firebase_performance: "^0.3.0+5"
```
If I then create a bundle and install it I see this in Logcat:

> 100-3100/? W/ComponentDiscovery: Class io.flutter.plugins.firebaseperformance.FlutterFirebaseAppRegistrar is not an found.
>     java.lang.ClassNotFoundException: io.flutter.plugins.firebaseperformance.FlutterFirebaseAppRegistrar
>         at java.lang.Class.classForName(Native Method)
>         at java.lang.Class.forName(Class.java:454)
>         at java.lang.Class.forName(Class.java:379)
>         at com.google.firebase.components.f.a(Unknown Source:29)
>         at com.google.firebase.components.f.a(Unknown Source:8)
>         at com.google.firebase.FirebaseApp.<init>(Unknown Source:76)
>         at com.google.firebase.FirebaseApp.a(Unknown Source:64)
>         at com.google.firebase.FirebaseApp.a(Unknown Source:2)
>         at com.google.firebase.FirebaseApp.b(Unknown Source:35)
>         at com.google.firebase.provider.FirebaseInitProvider.onCreate(Unknown Source:4)
>         at android.content.ContentProvider.attachInfo(ContentProvider.java:2092)
>         at android.content.ContentProvider.attachInfo(ContentProvider.java:2066)
>         at com.google.firebase.provider.FirebaseInitProvider.attachInfo(Unknown Source:3)
>         at android.app.ActivityThread.installProvider(ActivityThread.java:6983)
>         at android.app.ActivityThread.installContentProviders(ActivityThread.java:6528)
>         at android.app.ActivityThread.handleBindApplication(ActivityThread.java:6445)
>         at android.app.ActivityThread.access$1300(ActivityThread.java:219)
>         at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1859)
>         at android.os.Handler.dispatchMessage(Handler.java:107)
>         at android.os.Looper.loop(Looper.java:214)
>         at android.app.ActivityThread.main(ActivityThread.java:7356)
>         at java.lang.reflect.Method.invoke(Native Method)
>         at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
>         at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)
> 

But the app didn't crash.

So then I added the plugin which could provide the crash stacktrace from @remonh87 , and then indeed the app crashes when using bundles. Took me some time to figure out and setup the correct Proguard file.

@blasten The way I create and install bundles while testing was:
- Run `flutter build appbundle`
- Use the `bundletool` from Android to create an `apks` file: `bundletool build-apks --bundle=app.aab --output=app.apks`
- Use the `bundletool` to install it to an attached phone: `bundletool install-apks --apks=app.apks`

**Comment by blasten** [comment]
@bparrishMines, @collinjackson  Is `FlutterFirebaseAppRegistrar` instantiated by Firebase based on the `meta-data` in AndroidManifest.xml?

These classes will need the `@Keep` annotation, so they don't get removed by R8.

**Comment by collinjackson** [comment]
@bl

> @bparrishMines, @collinjackson Is `FlutterFirebaseAppRegistrar` instantiated by Firebase based on the `meta-data` in AndroidManifest.xml?
> 
> These classes will need the `@Keep` annotation, so they don't get removed by R8.

It sounds like this is actually kind of a Flutterfire issue. We can add the @Keep annotation there.

Is anyone seeing this issue who isn't using Flutterfire?

**Comment by cbracken** [comment]
Removing the engine label since this looks like it's build/plugin-specific.

**Comment by zanderso** [comment]
Adding the "waiting for customer response" label to keep this open to see if this has any other instances  beyond FlutterFire.

**Comment by PieterAelse** [comment]
From my side there are no other issues than the FlutterFirebaseAppRegistrar already mentioned.
Or are you waiting on response from @kroikie ?

**Comment by kroikie** [comment]
From the FlutterFire side this [PR](https://github.com/FirebaseExtended/flutterfire/pull/1373) should resolve the Registrar issue. If there is no issue on the Flutter side then I think this issue can be closed FlutterFire side is tracked [here](https://github.com/FirebaseExtended/flutterfire/issues/1337).

**Comment by zanderso** [comment]
Closing in favor of the FlutterFire issue.
